pushsync/pusher.go

```diff
@@ -0,0 +1,350 @@
+			}
+			return
+		}
+	}
+}
+
+// handleReceiptMsg is a handler for pssReceiptTopic that
+// - deserialises receiptMsg and
+// - sends the receipted address on a channel
+func (p *Pusher) handleReceiptMsg(msg []byte) error {
+	receipt, err := decodeReceiptMsg(msg)
+	if err != nil {
+		return err
+	}
+	p.logger.Trace("Handler", "receipt", label(receipt.Addr))
+	p.pushReceipt(receipt.Addr)
+	return nil
+}
+
+// pushReceipt just inserts the address into the channel
+func (p *Pusher) pushReceipt(addr []byte) {
+	select {
+	case p.receipts <- addr:
+	case <-p.quit:
+	}
+}
+
+// sendChunkMsg sends chunks to their destination
+// using the PubSub interface Send method (e.g., pss neighbourhood addressing)
+func (p *Pusher) sendChunkMsg(ch chunk.Chunk) error {
+	rlpTimer := time.Now()
+
+	cmsg := &chunkMsg{
+		Origin: p.ps.BaseAddr(),
+		Addr:   ch.Address()[:],
+		Data:   ch.Data(),
+		Nonce:  newNonce(),
+	}
+	msg, err := rlp.EncodeToBytes(cmsg)
+	if err != nil {
+		return err
+	}
+	p.logger.Trace("send chunk", "addr", label(ch.Address()))
+
+	metrics.GetOrRegisterResettingTimer("pusher.send.chunk.rlp", nil).UpdateSince(rlpTimer)
+
+	defer metrics.GetOrRegisterResettingTimer("pusher.send.chunk.pss", nil).UpdateSince(time.Now())
+	return p.ps.Send(ch.Address()[:], pssChunkTopic, msg)
+}
+
+// needToSync checks if a chunk needs to be push-synced:
+// * if not sent yet OR
+// * if sent but more than retryInterval ago, so need resend OR
+// * if self is closest node to chunk TODO: and not light node
+//   in this case send receipt to self to trigger synced state on chunk
+func (p *Pusher) needToSync(ch chunk.Chunk) bool {
+	item, found := p.pushed[ch.Address().Hex()]
+	now := time.Now()
+	// has been pushed already
+	if found {
+		// has synced already since subscribe called
+		if item.synced {
+			return false
+		}
+		item.lastSentAt = now
+	} else {
+		// first time encountered
+		addr := ch.Address()
+		hexaddr := addr.Hex()
+		// remember item
+		tag, _ := p.tags.Get(ch.TagID())
+		item = &pushedItem{
+			tag:         tag,
+			firstSentAt: now,
+			lastSentAt:  now,
+		}
+
+		// increment SENT count on tag  if it exists
+		if tag != nil {
+			tag.Inc(chunk.StateSent)
+			// opentracing for chunk roundtrip
+			_, span := spancontext.StartSpan(tag.Context(), "chunk.sent")
+			span.LogFields(olog.String("ref", hexaddr))
+			span.SetTag("addr", hexaddr)
+			item.span = span
+		}
+
+		// remember the item
+		p.pushed[hexaddr] = item
+		if p.ps.IsClosestTo(addr) {
+			p.logger.Trace("self is closest to ref: push receipt locally", "ref", hexaddr)
+			item.shortcut = true
+			go p.pushReceipt(addr)
+			return false
+		}
+		p.logger.Trace("self is not the closest to ref: send chunk to neighbourhood", "ref", hexaddr)
+	}
+	return true
+}
+
+// updateRetryInterval calibrates the period after which push index iterator restart from the beginning
+func (p *Pusher) updateRetryInterval(item *pushedItem, measurements uint64, average uint64) (uint64, uint64) {
```

Comment:
i don't see an upper bound here for the measurement. can this not create a situation where slow peers (or the current peer which is slow due to a slow connection) actually grind the push sync feature into a halt as a result?

Reply:
outliers are ignored that is the upper bound

Reply:
how and where are they ignored